pynativelib-proposal.rst

```diff
@@ -0,0 +1,797 @@
+    the various other representations that we care about.
+
+- Most library build systems won't have a flag that tells them to
+  generate a library called ``libpynativelib_openssl__ssl`` instead of
+  ``libssl`` or whatever. So we need a way to rename
+  libraries. Strategy:
+
+  - On Linux, .so files do know what their name is and this is
+    crucial, because if two libraries both think they have the same
+    name then you hit `glibc bug #19884
+    <https://sourceware.org/bugzilla/show_bug.cgi?id=19884>`_. However,
+    ``patchelf --set-soname`` lets you rename a shared library and
+    avoid this problem.
+
+  - On OS X, shared libraries do know what their name is, but I'm not
+    sure to what extent it affects things; however, my impression is
+    that their linker is weirdly concerned about such things
+    (FIXME). However, ``install_name_tool -id`` lets you change the
+    name of a shared library and avoid any potential problems.
+
+  - On Windows, .dll files do know what their name is, but I don't
+    *think* that anything cares? FIXME: we should check this. (If they
+    do care, then we can easily teach `redll
+    <https://github.com/njsmith/redll>`_ to modify the embedded name
+    strings, similar to ``patchelf --set-soname``.)
+
+    A bigger problem though is that on Windows, you don't link
+    directly against .dll files; instead, you pass the linker a .lib
+    "import library" and the .lib file says which .dll to link to. So
+    if you rename a .dll, you also need some way to modify all the
+    references to it inside the .lib file.
+
+    *Possible strategy 1:* rename the .dll files and then regenerate
+     the .lib files using something like `dlltool
+     <https://sourceware.org/binutils/docs/binutils/dlltool.html>`_.
+
+    *Possible strategy 2:* rename the .dll files and then patch the
+     .lib file to refer to the new name.
+
+    *Possible strategy 3:* leave the .lib file alone, and patch the
+     resulting binary using something like `redll
+     <https://github.com/njsmith/redll>`_.
+
+    Complicating strategies 1 & 2 is that there are actually two
+    different formats for .lib files -- the simple format that MSVC
+    traditionally uses, and the .a format traditionally used by
+    mingw-based compilers. The .lib format is very simple and it would
+    be easy to write a tool to patch it; also, it is the more
+    widespread format and mingw-based compilers can nowadays
+    understand it perfectly well, so that's probably what we want to
+    be standardizing on. Also, dlltool only knows how to generate the
+    .a format. Given all this, and the complexity/uncertainty
+    associated of post-hoc patching (strategy 3), I'm leaning towards
+    strategy 2: require everyone to use MSVC-style .lib files and
+    write a tool to patch them.
+
+    This also requires some way to actually generate .lib files when
+    using the mingwpy buildchain. This could use the new ``genlib``
```

Comment:
Unfortunately it's not quite so simple :-). You're right that the extension
doesn't mean anything, but nonetheless there are actually two different
formats in play. I'm not sure about binutils ld, but dlltool definitely
generates .a files that contain a bunch of COFF files (basically full
fledged .o files). MSVC .lib files OTOH are .a files that contain a bunch
of "ILF" format files. This is a very simple binary record format that just
names the relevant symbol and which dll it's in, without all the complexity
and overhead of generating real COFF. This is why msvc .lib files are
generally an order of magnitude smaller than binutils .a files. You can
find the gory details in the PE/COFF spec, which is freely available and
not too bad once you get oriented a bit.
.
Historically binutils had a nasty bug in reading ILF files, but we just
fixed that in binutils master as part of the mingwpy work. So when linking,
the two formats are functionally identical (except that I don't know how
well msvc handles the .a files -- but in theory it should work). But it
makes a non-trivial difference if we're talking about writing new code to
massage these files.
On Apr 17, 2016 6:27 PM, "Tony Kelman" <notifications@github.com> wrote:

> In pynativelib-proposal.rst
> <https://github.com/pypa/wheel-builders/pull/2#discussion_r59997934>:
>
> > +
> > +    Complicating strategies 1 & 2 is that there are actually two
> > +    different formats for .lib files -- the simple format that MSVC
> > +    traditionally uses, and the .a format traditionally used by
> > +    mingw-based compilers. The .lib format is very simple and it would
> > +    be easy to write a tool to patch it; also, it is the more
> > +    widespread format and mingw-based compilers can nowadays
> > +    understand it perfectly well, so that's probably what we want to
> > +    be standardizing on. Also, dlltool only knows how to generate the
> > +    .a format. Given all this, and the complexity/uncertainty
> > +    associated of post-hoc patching (strategy 3), I'm leaning towards
> > +    strategy 2: require everyone to use MSVC-style .lib files and
> > +    write a tool to patch them.
> > +
> > +    This also requires some way to actually generate .lib files when
> > +    using the mingwpy buildchain. This could use the new ``genlib``
>
> I think they're entirely equivalent and only differ in naming convention,
> mingw uses .dll.a for import libraries and .a for staric libraries just so
> you aren't calling both .lib. MSVC can use .dll.a import libs so I'm pretty
> sure the format is the same?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/pypa/wheel-builders/pull/2/files/f9da350f65be9bb7006cfa3c7388e19d1727a362#r59997934>
>